client/comms/wsconn.go

```diff
@@ -0,0 +1,275 @@
+	return NewWsConnWithPingWait(host, tlsCfg, pingWait)
+}
+
+// fetchConnected returns the connection connected state.
+func (conn *WsConn) fetchConnected() bool {
+	conn.connectedMtx.RLock()
+	connected := conn.connected
+	conn.connectedMtx.RUnlock()
+
+	return connected
+}
+
+// setConnected updates the connection's connected state
+func (conn *WsConn) setConnected(connected bool) {
+	conn.connectedMtx.Lock()
+	conn.connected = connected
+	conn.connectedMtx.Unlock()
+}
+
+// logRoute logs a request route as an id/route pair.
+func (conn *WsConn) logRoute(id uint64, route string) {
+	conn.reqMtx.Lock()
+	conn.req[id] = route
+	conn.reqMtx.Unlock()
+}
+
+// FetchRoute fetches the route of the provided request id.
+func (conn *WsConn) FetchRoute(id uint64) (string, error) {
+	conn.reqMtx.RLock()
+	route := conn.req[id]
+	conn.reqMtx.RUnlock()
+
+	if route == "" {
+		return "", fmt.Errorf("no route found for id #%v", id)
+	}
+
+	conn.reqMtx.Lock()
+	delete(conn.req, id)
+	conn.reqMtx.Unlock()
+
+	return route, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	url := url.URL{
+		Scheme: "ws",
+		Host:   conn.host,
+		Path:   "ws",
+	}
+
+	var err error
+	if conn.tlsCfg != nil {
+		dialer := &websocket.Dialer{
+			TLSClientConfig: conn.tlsCfg,
+		}
+		conn.ws, _, err = dialer.Dial(url.String(), nil)
+		if err != nil {
+			return err
+		}
+	} else {
+		conn.ws, _, err = websocket.DefaultDialer.Dial(url.String(), nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	conn.setConnected(true)
+	conn.ws.SetPingHandler(func(data string) error {
+		log.Tracef("Received ping")
+
+		now := time.Now()
+		err := conn.ws.SetReadDeadline(
+			now.Add(time.Duration(conn.pingWait) * time.Second))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v\n", err)
+			return err
+		}
+
+		// Respond with a pong.
+		return conn.ws.WriteControl(websocket.PongMessage, []byte{}, time.Time{})
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			connected := conn.fetchConnected()
+			if !connected {
+				log.Trace("reconnecting client")
+				err := conn.connect()
+				if err != nil {
+					log.Errorf("unable to establish a connection: %v\n", err)
+				}
+			}
+
+		case <-ctx.Done():
+			conn.ws.Close()
+			conn.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		// Only allow reads if there is an established connection.
```

Comment:
Why not put a select on ctx.Done here so you can shutdown this goroutine on command?

```go
select {
case <-ctx.Done():
	return
default:
}
```

Reply:
Regading https://github.com/decred/dcrdex/pull/55/files/d2c42302f14f561b7264fc5b3f8196a8a928f23d#r339298295

Reply:
using the select snippet above will no guarantee processing context signals because `ReadMessage` which blocks will be called after the default block. The context signal can come through while `ReadMessage` is blocking.This would interfere with a clean shutdown. I've made some modifications that I'm yet to push. The goroutine will be terminated by a close message.

Reply:
Ah, I see what you mean.